monkestation/code/modules/reagents/chemistry/machinery/atmos_machine.dm

```diff
@@ -0,0 +1,172 @@
+	if(is_operational)
+		begin_processing()
+	AddComponent(/datum/component/simple_rotation, ROTATION_ALTCLICK | ROTATION_CLOCKWISE | ROTATION_COUNTERCLOCKWISE | ROTATION_VERBS, null, CALLBACK(src, .proc/can_be_rotated))
+
+/obj/machinery/atmos_machine/proc/can_be_rotated(mob/user,rotation_type)
+	return !anchored
+
+/obj/machinery/atmos_machine/update_icon()
+	if((!is_operational) || (!on) || (reagents.total_volume == 0))
+		if (panel_open)
+			icon_state = "smoke0-o"
+		else
+			icon_state = "smoke0"
+	else
+		icon_state = "smoke1"
+	return ..()
+
+/obj/machinery/atmos_machine/RefreshParts()
+	. = ..()
+	var/new_volume = REAGENTS_BASE_VOLUME
+	for(var/obj/item/stock_parts/matter_bin/installed_bin in component_parts)
+		new_volume += REAGENTS_BASE_VOLUME * installed_bin.rating
+	if(!reagents)
+		create_reagents(new_volume)
+	reagents.maximum_volume = new_volume
+	if(new_volume < reagents.total_volume)
+		reagents.reaction(loc, TOUCH) // if someone manages to downgrade it without deconstructing
+		reagents.clear_reagents()
+	efficiency = 9
+	for(var/obj/item/stock_parts/capacitor/installed_cap in component_parts)
+		efficiency += installed_cap.rating
+	max_range = 1
+	for(var/obj/item/stock_parts/manipulator/installed_manip in component_parts)
+		max_range += installed_manip.rating
+	max_range = max(3, max_range)
+	multiplier = 0
+	for(var/obj/item/stock_parts/capacitor/installed_cap in component_parts)
+		multiplier += installed_cap.rating
+
+/obj/machinery/atmos_machine/on_set_is_operational(old_value)
+	if(old_value) //Turned off
+		end_processing()
+	else //Turned on
+		begin_processing()
+
+/obj/machinery/atmos_machine/process()
+	..()
+	if(reagents.total_volume == 0)
+		on = FALSE
+		update_icon()
+		return
+	if(on)
+		update_icon()
+		var/datum/reagents/chemholder = new(1000)
+		reagents.trans_to(chemholder, ((setting * 3) * 16) / efficiency)
+
+		for(var/datum/reagent/contained_reagent in chemholder.reagent_list)
+			var/temp = chemholder?.chem_temp
+			var/turf/turf = get_turf(src.loc)
+			if(!temp)
+				if(isopenturf(turf))
+					var/turf/open/open_turf = turf
+					var/datum/gas_mixture/air = open_turf.return_air()
+					temp = air.return_temperature()
+				else
+					temp = T20C
+			turf.atmos_spawn_air("[contained_reagent.get_gas()]=[(contained_reagent.volume * multiplier) /contained_reagent.molarity];TEMP=[temp]") //yes yes i know this is more chemicals than was inputed but like this would be slow as fuck otherwise
+			reagents.reagent_list -= contained_reagent
+		qdel(chemholder)
+
+/obj/machinery/atmos_machine/attackby(obj/item/I, mob/user, params)
+	add_fingerprint(user)
+	if(istype(I, /obj/item/reagent_containers) && I.is_open_container())
+		var/obj/item/reagent_containers/RC = I
+		var/units = RC.reagents.trans_to(src, RC.amount_per_transfer_from_this, transfered_by = user)
+		if(units)
+			to_chat(user, "<span class='notice'>You transfer [units] units of the solution to [src].</span>")
+			return
```

Comment:
Need an log_game or log_combat here, to catch turning the machine on and then filling it.